libhb/chroma_smooth.c

```diff
@@ -0,0 +1,370 @@
+
+        ctx->amount    = ctx->strength * 65536.0;
+        ctx->steps     = ctx->size / 2;
+        ctx->scalebits = ctx->steps * 4;
+        ctx->halfscale = 1 << (ctx->scalebits - 1);
+    }
+
+    if (chroma_smooth_init_thread(filter, 1) < 0)
+    {
+        chroma_smooth_close(filter);
+        return -1;
+    }
+
+    return 0;
+}
+
+static void chroma_smooth_thread_close(hb_filter_private_t *pv)
+{
+    int c, z;
+    for (c = 0; c < 3; c++)
+    {
+        chroma_smooth_plane_context_t * ctx = &pv->plane_ctx[c];
+        for (int t = 0; t < pv->threads; t++)
+        {
+            chroma_smooth_thread_context_t * tctx = &pv->thread_ctx[t][c];
+            for (z = 0; z < 2 * ctx->steps; z++)
+            {
+                free(tctx->SC[z]);
+                tctx->SC[z] = NULL;
+            }
+        }
+    }
+    free(pv->thread_ctx);
+}
+
+static int chroma_smooth_init_thread(hb_filter_object_t *filter, int threads)
+{
+    hb_filter_private_t * pv = filter->private_data;
+
+    chroma_smooth_thread_close(pv);
+    pv->thread_ctx = calloc(threads, sizeof(chroma_smooth_thread_context3_t));
+    pv->threads = threads;
+    for (int c = 0; c < 3; c++)
+    {
+        chroma_smooth_plane_context_t * ctx = &pv->plane_ctx[c];
+        int w = hb_image_width(ctx->pix_fmt, ctx->width, c);
+
+        for (int t = 0; t < threads; t++)
+        {
+            chroma_smooth_thread_context_t * tctx = &pv->thread_ctx[t][c];
+            int z;
+            for (z = 0; z < 2 * ctx->steps; z++)
+            {
+                tctx->SC[z] = malloc(sizeof(*(tctx->SC[z])) *
+                                     (w + 2 * ctx->steps));
+                if (tctx->SC[z] == NULL)
+                {
+                    hb_error("Chroma Smooth calloc failed");
+                    chroma_smooth_close(filter);
+                    return -1;
+                }
+            }
+        }
+    }
+    return 0;
+}
+
+static void chroma_smooth_close(hb_filter_object_t * filter)
+{
+    hb_filter_private_t *pv = filter->private_data;
+
+    if (pv == NULL)
+    {
+        return;
+    }
+
+    chroma_smooth_thread_close(pv);
+    free(pv);
+    filter->private_data = NULL;
+}
+
+static int chroma_smooth_work_thread(hb_filter_object_t *filter,
+                                     hb_buffer_t ** buf_in,
+                                     hb_buffer_t ** buf_out, int thread)
+{
+    hb_filter_private_t *pv = filter->private_data;
+    hb_buffer_t *in = *buf_in, *out;
+
+    if (in->s.flags & HB_BUF_FLAG_EOF)
+    {
+        *buf_out = in;
+        *buf_in = NULL;
+        return HB_FILTER_DONE;
+    }
+
+    out = hb_frame_buffer_init(in->f.fmt, in->f.width, in->f.height);
+
+    int c;
+    for (c = 0; c < 3; c++)
+    {
+        chroma_smooth_plane_context_t  * ctx  = &pv->plane_ctx[c];
+        chroma_smooth_thread_context_t * tctx = &pv->thread_ctx[thread][c];
+
+        if (c == 0)
+        {
+            // Luma
+            luma_copy(in->plane[c].data,
```

Comment:
If we typdefed the buffer plane struct instead of inlining it's definition in hb_buffer_t, this could be ```luma_copy(&in->plane[c], &out->plane[c])```.  Doing it this way reduces parameter count and has the additional benefit that source and dest with different width, height, or stride could be handled without changing the parameter list.

Reply:
Good idea.

Reply:
Let's leave this for when we overhaul all the filters this way.